### ISSUE TITLE
transfers not done in the curl_multi_add_handle() order

### ISSUE THREAD

**Issue by drok** [curl/curl#2701]
My application uses libcurl mult interface with an external libuv event loop. It needs to POST requests to a server in a pipelined manner, over a single connection, so that the requests arrive in the same order as they are sent. The write callback (CURLOPT_WRITEFUNCTION) sometimes adds new easy handles to be fetched with curl_multi_add_handle().

I find that often, when the write callback issues a new easy handle, while the internal libcurl queue/pipeline is already full, the newly added handle is executed ahead of the other requests already in the pipeline.

This is **not a pipelining bug**. See [update](https://github.com/curl/curl/issues/2701#issuecomment-408717419).

### I expected the following

Easy handles should be executed in the same order as they are submitted to libcurl with 

### curl/libcurl version

curl-7_58_0 (git)

### operating system

CentOS 6

I have created a testcase (test1950) which reproduces this problem. It reproduces it frequently (more than 90% of the time), but does not reproduce it at all when running under the debugger (tests/runtests.pl -g)

Unfortunately I'm not familiar enough with the curl codebase to suggest a fix at this time, but intend to try.

**Comment by bagder** [comment]
Pipelining is often horribly timing dependent and sensitive which makes it really annoying to debug.

If you're at all able to, I strongly advice you to switch to HTTP/2 multiplexing instead which is a much better use of a single connection and more reliable.

Do you know if this bug is a regression that used to work better before? Any particular reason you're testing 7.58.0 and not the latest?

**Comment by drok** [comment]
I can't use HTTP/2 unfortunately.

I've also tested 7.45.0, 7.50.0,  and 7.59.0. They all have similar behaviour wrt. pipeline ordering. 

I suspect the issue is related to scheduling newly added handles to be progressed asap via the splay tree (Curl_expire() call in curl_multi_add_handle()), so a handle added just as a head-of-pipe becomes DONE, is progressed before the pipeline is advanced via the Curl_splaygetbest() do-loop in multi-socket. Maybe the pipeline should be advanced as soon as a handle enters the DONE state, and then the splay tree be consulted.

I have also tried setting CURMOPT_PIPELINING to CURLPIPE_NOTHING, and setting the MAX_PIPELINE_LENGTH to a number larger than the total number of requests in the testcase, and neither had any effect.

**Comment by bagder** [comment]
> I have also tried setting CURMOPT_PIPELINING to CURLPIPE_NOTHING, ...  and neither had any effect.

Eh? You *disabled* pipelining and it still pipelined (in wrong order) ?

**Comment by bagder** [comment]
I presume you then simply mean that the HTTP requests weren't being issued in the same order the easy handles were added to the multi handle? Or was that the whole point from the start?

If so, I'd say that libcurl does in fact not promise to do that and it has no internal logic to keep operations ordered in such a manner. I'm not even sure I'm prepared to add that promise (by default) as I fear that risk complicating things...

**Comment by drok** [comment]
Trying to narrow down whether the issue is with any pipelining code, which may be tricky to debug, or somewhere else, I tried both enabling and disabling it. It's possible that with pipelining enabled, a request is sent to the server before the previous response is received (I haven't checked the wire protocol to confirm this), but it certainly does not affect the order of requests. But we can rule out pipelining logic being at fault here.

Generally, pipelining is supposed to add performance without changing behaviour (order of operations), and in this case I confirmed that it does not change behaviour (ie, maybe it does not need to be debugged).

The order of operations should be guaranteed, I think. Otherwise even harder to debug issues exist. Eg. Consider an application which in response to some random/external event issues two POSTs to the same URL. If one of the two submitted easy handles is always executed ahead of any previously queued requests, and the random events arrive at just the right frequency, the app can get into a state where only one POST gets sent to the server, jumping ahead of the queue, and all the previously queued requests are indefinitely delayed, pooling in an ever growing queue, and maybe even time out eventually. It would then be a nightmare to debug why some requests time out, while some are executed.

How would an application that requires order-of-execution guarantee be designed to work correctly? It would have to queue all non-idempotent requests itself, only submit them to curl one at a time, and only submit one if no other such request is still in-flight.

If you think of http POSTs as normal writes to a file, you'd expect that sequential writes get written in the same order as posted, regardless of the timing of each individual write. Would it be reasonable that for some timing of write()'s, the data appears in the file in a different order than the order it was written?

How would you suggest I code my application to guarantee the order that requests arrive at the server? Let's say the data being posted is temperature measurements, such that together, all the measurements are a time-series. The server should just append it to some file as it's received, not try to reorder arriving requests before writing it (which would unnecessarily consume memory, and would require much more complex server-side file-writing logic).

**Comment by drok** [comment]
Having looked into this for a couple of days now, I think the code **is** designed with request  ordering in mind. Incoming requests are placed in a queue, and are processed in the same order. I think the problem is that there is an additional queue, the splay tree, and in multi_socket(), Curl_splaygetbest() is used opportunistically to handle requests based on their connection/timeout status. If the splay tree were used with a bit of care wrt **handle state**, then execution order can be guaranteed.

My application would have to be redesigned without this guarantee, and I feel it would be less effort if I just fixed libcurl instead. Let me look into it for a few days, I think the solution may be quite simple.

To answer my own question, I think the redesign would mean that I'd need a separate worker thread for libcurl managing its own sockets and event loop, while the main thread would handle all the other event sources with a libuv or such. Currently the application is designed with a single event loop, and single thread. It works great, except when a random event adds a new easy handle to curl's queue while curl has just finished another request.

With respect to the documentation; which ordering is not guaranteed, it is **provided** when libcurl manages its own event loop, in my experience. It's also **provided** when libcurl is using an external event loop, **if** all the requests are delivered to libcurl _synchronously_, eg. from within the CURLOPT_WRITEFUNCTION callback.

The order is **not provided** when libcurl is using an external event loop, **and** curl_multi_add_handle() is called asynchronously, ie, when libcurl is waiting for socket activity.

**Comment by drok** [comment]
Conversely, if the "**ordering is not guaranteed**" language were added to the documentation, it would then be ok to also add code to randomize libcurl's internal queue. A random queue (really not a "queue", but a "set", or pile) does conform to the non-guarantee.

A lot of libcurl's users would think the random order of execution is either a bug or an anti-feature, and would want it either fixed, or would look for alternates to libcurl.

If you were to say "not guaranteed to not have buffer overflows", the users' reaction would be similar.

I believe "guaranteed" does not mean "it executes in the _correct_ order" or that "it contains no buffer overflows", but rather that you're committed to treating either instance as an unintended error, and to fix it.

**Comment by bagder** [comment]
It is not a "random order". It is a "not guaranteed to be in the same order as the handles were added". I don't see how randomizing these things would improve the experience for users.

The ordering is basically attempted to be kept but I've never considered it to be important nor a property to test for/guarantee.

As for the *there is only one test case that uses CURLMOPT_SOCKETFUNCTION*, I think you're disregarding the `./runtests.pl -e` option which runs *all* tests via the event-based "simulation" in the debug build and thus verifying that using the event-based API works the same way as the non-event one. We run that on every commit and PR.

I wouldn't object to a fix that would make libcurl adhere to the ordering more/better (which for example could make the splay tree sort nodes expiring on the same millisecond in the order of creation, possibly by adding "resolution" by the means of an additional mere counter), but I would still hesitate to *guarantee* it. Again, especially when for example you set a limit to number of connections/requests per host/connection and libcurl queues them up internally. But it's not a very strong opinion of mine, if we can fix the ordering without it costing is too much in the form of code complexity or performance, I can be persuaded.

> If you were to say "not guaranteed to not have buffer overflows", the users' reaction would be similar.

**libcurl is not guaranteed to not have buffer overflows**. That would be preposterous given the fact that we've announced buffer overflows several times in the past and it would be strange to presume that we've fixed the last one now. I can guarantee that we do our very best to avoid them though.

**Comment by drok** [comment]
A random ordering would help developers who use libcurl to test their assumptions. In my case, I assumed libcurl always executes requests in the same order as submitted, as it appears to do. This assumption is incorrect, and I would have not designed the app by putting libcurl in the same event loop as my other code, but would have chosen to put it in a separate thread, and not use its external event features, which cause out-of-order execution. My application requires guaranteed order of execution, and I should have designed a wrapper around libcurl to provide this guarantee. Instead, I found out the hard and expensive way that this wrapper is required. As currently designed (one event loop), there are infrequent errors when requests get handled out-of-sequence, which makes it very hard to diagnose. If libcurl would randomize the request order, it would have been much easier to diagnose the errors and to draw the conclusion that ordering is not guaranteed.

That said, randomizing the order now is merely an academic thought exercise. Implementing one would break a lot of applications which simply assume requests ordering, and get lucky 99.9% of the time.

Now, most of the apps out there which use libcurl will at some time execute some spurious request sequence in the wrong order. The effects may vary, and the occurrence is likely infrequent enough that the unexpected behaviour probably gets blamed on 'gremlins'. 

Anyway, I am working on implementing guaranteed execution order. As it is written now, libcurl is not far off the mark, and I think it is less effort for me to fix libcurl than to redesign my app with some multi-threaded wrapper, locking, and dealing with the inevitable bugs.

Thanks for pointing out the `runtests.pl -e` option, I was not aware of it, though having looked into it, it's not clear to me what it does, and how it affects the API tests in `tests/libtest`.

Finally I think we are in agreement about what a **guarantee** means: not a certification, but rather a commitment to fix out-of-spec behaviour when discovered.

**Comment by bagder** [comment]
libcurl still behaves *the same way* when run again under the exact same circumstances. That's quite a different thing than *randomizing* how it behaves.

> I am working on implementing guaranteed execution order

Just be careful so that you don't overpromise on that guarantee. There's a lot of different network stuff going on, with queues and multiplexing etc, so the specific wording of this guarantee will be important. I'll of course be happy to review/merge any work as soon as you have something to show and test.

> Thanks for pointing out the runtests.pl -e option, I was not aware of it, though having looked into it, it's not clear to me what it does, and how it affects the API tests in tests/libtest.

It is primarily a way to run all "normal" tests and make sure that using the event-based API works out exactly the same way as otherwise, so only tests using the command line "curl" are affected at this point. We *could* make all tests using `curl_easy_perform()` do it, but I haven't done the necessary work for the libtests to go that way.

**Comment by bagder** [comment]
First. Beware that HTTP/1.1 [pipelining is marked for deprecation](https://curl.haxx.se/dev/deprecate.html). There's no point in doing much work on pipelining now. In fact, once we get rid of pipelining once and for all, the code will get a chance to improve significantly since we can remove a large amount of special-case code.

Then, note that I fixed several "confusions" recently in master where "job" and "connection" data was conflicted or modified when it shouldn't. I'm not convinced it is entirely cleared up yet but we're getting there, and that's the primary reason for the planned pending 7.61.1 release.

I get your desire to create a "job" object but can you clarify how that is not the same as `struct SingleRequest` within the `Curl_Easy` ? An easy handle represents a single transfer and I can't see that a "job" would be much different. (Even if we of course always can improve the code.)

I can't quite understand your resentment for `EXPIRE_NOW` since it is there basically due to how the multi API works - we can only ask for "action" based on time or socket activity so unless you want it time based I presume you're thinking of using an internal socket? (unless you change behavior of the API?)

Whatever rewrite you're planning, I hope yo plan on doing it with small separate steps at a time so that we can discuss and review them as you go along, to avoid a huge undertaking that we can't accept later on!

**Comment by drok** [comment]
It makes me very sad that instead of fixing the pipeline implementation, you've decided to remove it. It will of course, remain part of the HTTP/1.1, ESMTP, etc specifications, as well as server codebases.

My plan with respect to the multi.c rewrite is to make pipelining an integral, first-class citizen, not a special case. In light of your planned deprecation of pipelining, this means my multi.c will be of little value to your project, and that I will likely continue with a fork of libcurl that fully implements pipelining and guaranteed execution order, for my project. My application requires it for performance, and libcurl sans pipelining would be a poor choice for my project, and one I would be stuck with, as there aren't good alternatives to libcurl at the moment.

> I get your desire to create a "job" object but can you clarify how that is not the same as struct SingleRequest within the Curl_Easy ?

The SingleRequest struct appears to be buffer state. To me, a "job" is the URL that libcurl is asked to fetch when the application submits it by calling `curl_multi_add_handle()`. The job struct should contain job configuration (URL, headers, compression, TLS options, etc) as well as protocol and buffer state, so the SingleRequest struct is only a small part of the 'job'.

> we can only ask for "action" based on time or socket activity

Actually, in the event-loop paradigm, the only blocking point should be the event loop itself. For a healthy application, the only requirement of the application code and libraries is that it not block. So, scheduling an immediate callback for something that can happen 'now' is as unnecessary, as it is awkward to describe. Instead, when a new handle is added, I think libcurl should do all the processing it needs to up until the next blocking call, then pass control back to the event loop. This would include finding the right connection for the job, opening sockets if necessary, What's the point in postponing non-blocking work?

Speaking of the event-loop paradigm, the filesystem access that libcurl needs, such as reading certificates, etc, should also use the event loop for the blocking bits. This may be the subject of a future issue. My application is not much affected by the blocking that libcurl does by using old-school file access for these files, so I don't really care about it.

The real problem is that this `EXPIRE_NOW` way of doing things "now-ish" allows the race condition that manifests as out-of-order URL fetching. If you submit a job with `curl_multi_add_handle()` from within the `curl_multi_socket_action()` call, ie, from a libcurl callback, the job gets scheduled in the expected sequence, at the end of the queue. If however, the job is submitted from outside the `curl_multi_socket_action()` call, ie, from another event loop callback, then depending on the state of libcurl's queue, the new job may be executed ahead of other previously queued jobs.

You suggested adding extra resolution to the splay tree to try to keep the ordering, and it may even work, but it's a hack that will likely break, as like you said, few other developers are affected by out-of-order execution, so when it breaks, it would not be noticed for a long time.

The most relevant question here should be; since my main drive is pipelining and guaranteed order execution, and you're planning on deprecating pipelining, what value is there to me contributing my work to your project? I think you should see my initiative as an opportunity to cancel your deprecating plans, and adopt my changes instead. I'm not certain how small the steps can be, given that I will be reorganizing some of the internal data structures. At least, my changes will be contained to mostly multi.c, maybe that is 'small' enough to discuss and review. Based on my initial progress, I'll be removing more code than I'm adding.

**Comment by drok** [comment]
**Update**. I understand this bug and the fix required much better now.

It is **not a pipelining bug**, as disabling pipelining does not fix it. The bug is that curl sends requests to the server in a different order than the application submitted the respective easy handles, and it occurs when all the following are true:

- libcurl uses an external event loop (CURLMOPT_SOCKETFUNCTION)
- jobs are submitted to curl from outside the call stack of `curl_multi_socket_action()` (ie, not from a libcurl socket callback)
- there is at least one active transfer and one pending transfer, when the a new job is submitted.
- `CURLMOPT_MAX_HOST_CONNECTIONS = 1`, ie, connection reuse is active, all requests go over the same connection.
- The currently active job must be ready to transition to the `DONE` state on the next call to `curl_multi_socket_action()` following the new job submission.

The out-of-order execution happens as follows:

When a new job is submitted with `curl_multi_add_handle()`,  it is appended to the `pending` list, and an  `EXPIRE_NOW` timer that is added to the splay tree. When the next socket to be processed causes the active job's FSM to transition to the `DONE` state, `multi_socket()` picks the first job on the splay tree, which is the most recently added job, and tries to progress it to the CONNECT state. As the previous job is `DONE`, this most recently added job is successfully progressed to `CONNECT`, ahead of the job that was already waiting on the pending list. On the other hand, if the currently active job is not quite ready to transition to `DONE` before the next splay tree scan, the newly added job will remain at the back of the `pending` list, and will be executed in order, after the previously pending job.

In conclusion, it is the `EXPIRE_NOW` logic that inserts newly submitted jobs ahead of previously pending ones. This is logic makes newly submitted jobs be considered **as soon as possible**, rather than **after the other pending jobs**

**Comment by bagder** [comment]
> The bug is that curl sends requests to the server in a different order than the application submitted the respective easy handles

I feel a need to stress that **this is not a bug**. Nowhere does libcurl guarantee this order. Never did.

Guaranteeing the order will even be really hard and will require a set of conditions (like the same connection must be used etc) which an application typically has a hard time to control or even be aware of.
Simply starting the transfers in the add_handle() order will certainly not be enough to guarantee that they will eventually "happen" (whatever definition of "happen" that we choose) in that order.

Maybe we can start here. What exactly would you expect or want libcurl to guarantee here?

**Comment by drok** [comment]
Then it should document that the order of execution may be different than the order of job submission, even for simple cases like all the requests are destined for a single server, and then only one connection is allowed.

I would not have wasted so much time using the wrong library for my project. Likewise for other developers who've encountered this problem before, and with whom you've discussed this guarantee before (eg, Cristian Morales Vega, circa Jun 18, 2015, at 7.43.0, maybe others too).

It's a shame too, because the only reason requests are processed out of order is the EXPIRE_NOW scheme of postponing until "soonish" work that can be performed immediately, because unspecified reasons.

**Comment by bagder** [comment]
I'll just repeat myself:

**What exactly would you expect or want libcurl to guarantee here?**

**Comment by drok** [comment]
**I expect libcurl to send requests to the server in the same order it receives them from my application.** ... in the simplest of operating modes, ie, one persistent HTTP/1.1 connection to one server, 3 measly requests, all POSTs to the same URL.

I expect the library to deliver the same guarantees that the underlying tech guarantees. Data to be sent to the server in the same order as my application sends it, and to responses delivered to my application in the same order as received from the server, while behaving as specified in one or more RFCs.

The purpose of a library is to ease my task of writing common code, by providing a **library** of functions that implement commonly needed code. In libcurl's case, as it claims to be a HTTP library, I would also expect the relevant RFCs to be implemented as they are published.

If I implemented my app from scratch, without a library, I would want one TCP connection, I'd want HTTP 1.1 spoken on this connection, with TLS encryption and zlib compression.

TCP guarantees data arrives in the same order as sent. I want this guarantee, which is promised by an RFC. I expect errors to be delivered to my app if they cannot be handled transparently. (disconnects, timeouts, etc). Providing ordered data **most of the time** is not good **any of the time**

TLS guarantees to check various info upon connection, and to encrypt the data a specific way, as specified by an RFC. I expect the same guarantees that the RFC makes. I expect errors to be communicated when a guarantee is not available (expired certs, unsupported cipher, etc). Best effort to implement the RFC is not good enough. Providing the expected cert checking or encryption strength **most of the time** is not good enough **any of the time**.

HTTP makes some guarantees too, specified in one or more RFCs. I want all of those. Not a best attempt at a rendition of the RFC.

If I were implementing the protocol without a library, my application would send as many requests to the server as it needs to, in the order that it needs to, and the server would respond to them in the same order (or close the connection). I don't see why I should give up this basic feature just because I delegated the common tasks of communicating with a server to a library whose core competency is allegedly, HTTP communication.

If I wanted an approximate rendition of the HTTP protocol, I would implement it myself, and live with some bugs. I delegated the task to a popular, and actively maintained library, exactly to avoid approximation. The assumption (wrong as it turns out) is that well maintained and well tested code implements standard protocols as specified by the relevant RFCs, not some casual best-effort attempt at maybe implementing just the easy part of that RFC, and providing an "almost always works" non-guarantee for the rest.
